File: rust/benches/latency_rtt.rs
//! Ping-pong round-trip latency, this crate vs rtrb.
//!
//! The main thread stamps a token into one ring and spins until an echo thread
//! bounces it back through a second ring; each round trip is recorded into an HDR
//! histogram. Both threads are pinned and busy-spin. Halve the figures for the
//! one-way hand-off. The per-round `Instant` pair adds a fixed timer overhead
//! (tens of ns) that is identical across contenders, so the comparison is fair.

use std::thread;
use std::time::Instant;

use hdrhistogram::Histogram;

const ROUNDS: u64 = 2_000_000;
const WARMUP: u64 = 100_000;
const PING_CORE: usize = 0;
const PONG_CORE: usize = 2;

fn pin(core: usize) {
    if let Some(ids) = core_affinity::get_core_ids() {
        if let Some(id) = ids.into_iter().find(|c| c.id == core) {
            core_affinity::set_for_current(id);
        }
    }
}

fn report(name: &str, hist: &Histogram<u64>) {
    // The mean is precise (quantization averages out over millions of samples);
    // the per-op percentiles are bucketed at the Windows timer granularity
    // (~100 ns) and so are only indicative until measured with a TSC clock.
    println!(
        "  {name:<12} mean {:>5.1} ns   p50 {:>4} ns   p99 {:>4} ns   max {:>6} ns",
        hist.mean(),
        hist.value_at_quantile(0.50),
        hist.value_at_quantile(0.99),
        hist.max(),
    );
}

fn bench_spsc_ring() -> Histogram<u64> {
    let (mut ping_tx, mut ping_rx) = spsc_ring::channel::<u64>(2);
    let (mut pong_tx, mut pong_rx) = spsc_ring::channel::<u64>(2);

    let echo = thread::spawn(move || {
        pin(PONG_CORE);
        loop {
            let value = loop {
                match ping_rx.pop() {
                    Some(v) => break v,
                    None => std::hint::spin_loop(),
                }
            };
            if value == u64::MAX {
                break;
            }
            while pong_tx.push(value).is_err() {
                std::hint::spin_loop();
            }
        }
    });

    pin(PING_CORE);
    let mut hist = Histogram::<u64>::new(3).unwrap();
    for i in 0..(WARMUP + ROUNDS) {
        let start = Instant::now();
        while ping_tx.push(1).is_err() {
            std::hint::spin_loop();
        }
        loop {
            if pong_rx.pop().is_some() {
                break;
            }
            std::hint::spin_loop();
        }
        if i >= WARMUP {
            hist.record(start.elapsed().as_nanos() as u64).unwrap();
        }
    }
    while ping_tx.push(u64::MAX).is_err() {
        std::hint::spin_loop();
    }
    echo.join().unwrap();
    hist
}

fn bench_rtrb() -> Histogram<u64> {
    let (mut ping_tx, mut ping_rx) = rtrb::RingBuffer::<u64>::new(2);
    let (mut pong_tx, mut pong_rx) = rtrb::RingBuffer::<u64>::new(2);

    let echo = thread::spawn(move || {
        pin(PONG_CORE);
        loop {
            let value = loop {
                match ping_rx.pop() {
                    Ok(v) => break v,
                    Err(_) => std::hint::spin_loop(),
                }
            };
            if value == u64::MAX {
                break;
            }
            while pong_tx.push(value).is_err() {
                std::hint::spin_loop();
            }
        }
    });

    pin(PING_CORE);
    let mut hist = Histogram::<u64>::new(3).unwrap();
    for i in 0..(WARMUP + ROUNDS) {
        let start = Instant::now();
        while ping_tx.push(1).is_err() {
            std::hint::spin_loop();
        }
        loop {
            if pong_rx.pop().is_ok() {
                break;
            }
            std::hint::spin_loop();
        }
        if i >= WARMUP {
            hist.record(start.elapsed().as_nanos() as u64).unwrap();
        }
    }
    while ping_tx.push(u64::MAX).is_err() {
        std::hint::spin_loop();
    }
    echo.join().unwrap();
    hist
}

fn main() {
    println!("ping-pong round trip ({ROUNDS} rounds, cores {PING_CORE}<->{PONG_CORE})");
    report("spsc-ring", &bench_spsc_ring());
    report("rtrb", &bench_rtrb());
}

File: cpp/bench/latency_rtt.cpp
// Ping-pong round-trip latency, this library vs rigtorp::SPSCQueue.
//
// The main thread stamps a token into one ring and spins until an echo thread
// bounces it back through a second ring; each round trip is recorded. Both
// threads are pinned and busy-spin. Halve the figures for the one-way hand-off.
// Timestamps come from tsc-latency's fenced RDTSC pair, so the percentiles are
// real (steady_clock on Windows quantizes at ~100 ns and only the mean survives).

#include <chrono>
#include <cstdint>
#include <cstdio>
#include <numeric>
#include <thread>
#include <vector>

#include <rigtorp/SPSCQueue.h>
#include <tsclat/clock.hpp>
#include <tsclat/tsc.hpp>

#include "bench_util.hpp"
#include "spsc/ring_buffer.hpp"

namespace {

constexpr int kRounds = 2'000'000;
constexpr int kWarmup = 100'000;
constexpr unsigned kPingCore = 0;
constexpr unsigned kPongCore = 2;

void report(const char* name, std::vector<std::uint64_t> ticks) {
    const auto& clock = tsclat::TscClock::instance();
    const double mean =
        clock.to_ns(std::accumulate(ticks.begin(), ticks.end(), std::uint64_t{0})) /
        static_cast<double>(ticks.size());
    std::sort(ticks.begin(), ticks.end());
    std::printf(
        "  %-12s mean %5.1f ns   p50 %4.0f ns   p99 %4.0f ns   p99.9 %5.0f ns   max %6.0f ns\n",
        name, mean, clock.to_ns(bench::quantile(ticks, 0.50)),
        clock.to_ns(bench::quantile(ticks, 0.99)), clock.to_ns(bench::quantile(ticks, 0.999)),
        clock.to_ns(ticks.back()));
}

std::vector<std::uint64_t> bench_spsc_ring() {
    auto to_echo = spsc::channel<std::uint64_t>(2);
    auto from_echo = spsc::channel<std::uint64_t>(2);
    auto ping_tx = std::move(to_echo.first);
    auto pong_rx = std::move(from_echo.second);

    std::thread echo([ping_rx = std::move(to_echo.second),
                      pong_tx = std::move(from_echo.first)]() mutable {
        bench::pin_to_core(kPongCore);
        std::uint64_t value = 0;
        for (;;) {
            while (!ping_rx.pop(value)) {
            }
            if (value == UINT64_MAX) {
                break;
            }
            while (!pong_tx.push(value)) {
            }
        }
    });

    bench::pin_to_core(kPingCore);
    std::vector<std::uint64_t> samples;
    samples.reserve(kRounds);
    std::uint64_t echoed = 0;
    for (int i = 0; i < kWarmup + kRounds; ++i) {
        const std::uint64_t start = tsclat::tsc_begin();
        while (!ping_tx.push(1)) {
        }
        while (!pong_rx.pop(echoed)) {
        }
        if (i >= kWarmup) {
            samples.push_back(tsclat::tsc_end() - start);
        }
    }
    while (!ping_tx.push(UINT64_MAX)) {
    }
    echo.join();
    return samples;
}

std::vector<std::uint64_t> bench_rigtorp() {
    rigtorp::SPSCQueue<std::uint64_t> to_echo(2);
    rigtorp::SPSCQueue<std::uint64_t> from_echo(2);
    std::thread echo([&] {
        bench::pin_to_core(kPongCore);
        for (;;) {
            while (to_echo.front() == nullptr) {
            }
            const std::uint64_t value = *to_echo.front();
            to_echo.pop();
            if (value == UINT64_MAX) {
                break;
            }
            while (!from_echo.try_push(value)) {
            }
        }
    });

    bench::pin_to_core(kPingCore);
    std::vector<std::uint64_t> samples;
    samples.reserve(kRounds);
    for (int i = 0; i < kWarmup + kRounds; ++i) {
        const std::uint64_t start = tsclat::tsc_begin();
        while (!to_echo.try_push(1)) {
        }
        while (from_echo.front() == nullptr) {
        }
        from_echo.pop();
        if (i >= kWarmup) {
            samples.push_back(tsclat::tsc_end() - start);
        }
    }
    while (!to_echo.try_push(UINT64_MAX)) {
    }
    echo.join();
    return samples;
}

}  // namespace

int main() {
    const auto& clock = tsclat::TscClock::instance();
    std::printf(
        "ping-pong round trip (%d rounds, cores %u<->%u, invariant tsc: %s, %.4f ns/tick)\n",
        kRounds, kPingCore, kPongCore, tsclat::has_invariant_tsc() ? "yes" : "no",
        clock.ns_per_tick());
    report("spsc::Ring", bench_spsc_ring());
    report("rigtorp::SPSCQueue", bench_rigtorp());
    return 0;
}
